rpcs3/Emu/Cell/Modules/cellSearch.cpp

```diff
@@ -98,17 +186,56 @@ error_code cellSearchStartListSearch(CellSearchListSearchType type, CellSearchSo
 		return CELL_SEARCH_ERROR_PARAM;
 	}
 
+	switch (type)
+	{
+		case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_ALBUM:
+		case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_GENRE:
+		case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_ARTIST:
+		case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_YEAR:
+		case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_MONTH:
+		case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_ALBUM:
+		case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_PLAYLIST:
+		case CELL_SEARCH_LISTSEARCHTYPE_VIDEO_ALBUM:
+		case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_PLAYLIST:
+			break;
+		default:
+			return CELL_SEARCH_ERROR_PARAM;
+	}
+
+	if (sortOrder != CELL_SEARCH_SORTORDER_ASCENDING && sortOrder != CELL_SEARCH_SORTORDER_DESCENDING)
+	{
+		return CELL_SEARCH_ERROR_PARAM;
+	}
+
 	const auto search = g_fxo->get<search_info>();
 
+	switch(search->state)
+	{
+		case SearchState::IDLE:
+			break;
+		case SearchState::NOT_INITIALIZED:
+		case SearchState::INITIALIZING:
+			return CELL_SEARCH_ERROR_NOT_INITIALIZED;
+		case SearchState::FINALIZING:
+			return CELL_SEARCH_ERROR_FINALIZING;
+		case SearchState::IN_PROGRESS:
+		case SearchState::CANCELING:
+			return CELL_SEARCH_ERROR_BUSY;
+		default:
+			return CELL_SEARCH_ERROR_GENERIC;
+	}
+
 	*outSearchId = idm::make<search_object_t>();
 
+	search->state = SearchState::IN_PROGRESS;
 	sysutil_register_cb([=](ppu_thread& ppu) -> s32
 	{
 		vm::var<CellSearchResultParam> resultParam;
 		resultParam->searchId = *outSearchId;
 		resultParam->resultNum = 0; // TODO
 
-		search->func(ppu, CELL_SEARCH_EVENT_LISTSEARCH_RESULT, CELL_OK, vm::cast(resultParam.addr()), search->userData);
+		search->state = SearchState::IDLE;
```

Comment:
Something is probably needed to done about thread-safety, especially since the callback is executed in another.

Reply:
Is concurrent execution of those possible? It technically is but `cellSearch` explicitly states it's not thread safe so if application calls methods without waiting for the completion callback `search->func` then I don't think it's our problem.

Reply:
Doesn't matter, emulation speeds != original hw speeds, race conditions which on original hw never bothered anyone may cause many problems with emulation.
We don't want to leave holes for such bugs.

Reply:
In other words lots of games are trolls and may not use thread-safe protection with cellSearch methods, and our code should take care of it as much as it can.